aesd-char-driver/main.c

```diff
@@ -0,0 +1,339 @@
+    filp->private_data = info_dev;
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    /**
+     * Remove assignment of private data.
+     */
+    filp->private_data = NULL;
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    struct aesd_dev *data;
+    size_t relative_byte_offset;
+    struct aesd_buffer_entry *ret_buf;
+    int bytes_to_eob;
+    int bytes_to_copy;
+    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+    if(!filp || !filp->private_data)
+    {
+        retval = -EINVAL;
+        goto ret_func;
+    }
+    data = filp->private_data;
+    //block on mutex lock, can be interrupted by signal
+    if (mutex_lock_interruptible(&data->lock))
+    {
+        retval = -EINTR;
+        
+        //Add a goto statement to the end
+        goto ret_func;
+    }
+    //find data at an offsert fpos
+    ret_buf = aesd_circular_buffer_find_entry_offset_for_fpos(&data->circular_buffer,
+                                                            *f_pos, 
+                                                            &relative_byte_offset);
+    if(!ret_buf)
+    {
+        //free MUTEX and exit with retval as 0 (EOF reached)
+        *f_pos = 0;
+        goto free_lock;
+    }
+    //bytes to end of buffer 
+    bytes_to_eob = (ret_buf->size - relative_byte_offset);
+    //return either upto end of buffer or count, whichever is minimum
+    bytes_to_copy = bytes_to_eob<count?bytes_to_eob:count;
+    //update fpos (partial read rule)
+    *f_pos += bytes_to_copy;
+    PDEBUG("Returng %s\n",ret_buf->buffptr);
+    //copy to user buffer
+    if (copy_to_user(buf, (ret_buf->buffptr + relative_byte_offset), bytes_to_copy)) 
+    {
+		retval = -EFAULT;
+		//free MUTEX and exit with retval as -EFAULT
+        goto free_lock;
+	}
+    //return number of bytes copied to buffer
+    retval += bytes_to_copy;
+    free_lock: mutex_unlock(&data->lock);
+    ret_func: return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    struct aesd_dev *data;
+    struct aesd_buffer_entry enqueue_buffer;
+    struct working_buffer copy_buffer;
+    int mem_to_malloc,i,start_ptr = 0;
+    char *free_buffer;
+    if(!filp || !filp->private_data)
+    {
+        retval = -EINVAL;
+        goto ret_func;
+    }
+    data = filp->private_data;
+
+    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+    if(count == 0)
+    {
+        goto ret_func;
+    }
+
+    //block on mutex lock, can be interrupted by signal
+    if (mutex_lock_interruptible(&data->lock))
+    {
+        retval = -EINTR;
+        
+        //Add a goto statement to the end
+        goto ret_func;
+    }
+    
+    //Allocate normal kernel ram. May sleep.
+    copy_buffer.buffer = kmalloc(count,GFP_KERNEL);
```

Comment:
@GuruprashanthKrishnakumar does this need to be inside the mutex lock?

Reply:
I guess I locked it very coarsely because the copy buffer is a local variable, so it should be fine across multiple simultaneous write calls. Would it be right to lock it just before starting to parse the copy buffer? Because once I start parsing the copy buffer I also manipulate the working buffer which is a global and hence is not reentrant across multiple write calls.